### PR TITLE
Implement _compute_intra_grad_corr_mean for gradient computation

### PR DESCRIPTION
**Pull request by cyugao**
## What does this PR do?
Implement `_compute_intra_grad_corr_mean` and tests
This utility function can be helpful for analyzing learning behavior.

## Before submitting

- [ ] Did you have fun?
  - Make sure you had fun coding 🙃
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/main/CONTRIBUTING.md)?
- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [ ] N/A
- [ ] Did you make sure to update the docs?
  - [ ] N/A
- [x] Did you write any new necessary tests?
  - [ ] N/A
- [ ] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/main/CHANGELOG.md)? (if needed)
  - [ ] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -120,7 +120,7 @@ install_dep_pytorch_nightly: &install_dep_pytorch_nightly
         # check if we have restored venv cache (/home/circleci/venv) correctly, if so, just skip
         if [ -f /home/circleci/venv/check_version.py ]; then python /home/circleci/venv/check_version.py torch eq 1.12 && exit 0; fi
         # start installing
-        pip install --pre torch==1.13.0.dev20220825+cu113 torchvision==0.14.0.dev20220825+cu113 --extra-index-url https://download.pytorch.org/whl/nightly/cu113
```
Comment (min-xu-ai) [inline .circleci/config.yml:123]: can you update this line to download the latest nightly, not the stable version?

File: .circleci/config.yml [hunk 2]
```diff
@@ -105,7 +105,7 @@ install_dep_pytorch_stable: &install_dep_pytorch_stable
         # check if we have restored venv cache (/home/circleci/venv) correctly, if so, just skip
         if [ -f /home/circleci/venv/check_version.py ]; then python /home/circleci/venv/check_version.py torch eq 1.11 && exit 0; fi
         # start installing
-        pip install --progress-bar off torch==1.12.0+cu113 torchvision==0.13.0+cu113 --extra-index-url https://download.pytorch.org/whl/cu113
+        pip install --progress-bar off torch==1.12.1+cu113 torchvision==0.13.1+cu113 --extra-index-url https://download.pytorch.org/whl/cu113
```
Comment (min-xu-ai) [inline .circleci/config.yml:108]: we should test the latest stable version, which is 1.13 here. Overall, we test 3 versions:

lts: 1.8.x
stable: 1.13.x now
nightly: whatever next version

File: fairscale/optim/adascale.py [hunk 3]
```diff
@@ -387,6 +387,44 @@ def _update_avg(self, name: str, value: np.ndarray, factor: float) -> None:
             else:
                 self._state[name] = factor * self._state[name] + (1.0 - factor) * value
 
+    def _gather_flat_grad(self) -> torch.Tensor:
+        """
+        Helper function for gathering all gradients into a single vector.
+        Duplicated from torch.optim.lbfgs.
+        """
+        views = []
+        for param_group in self._optimizer.param_groups:
```
Comment (min-xu-ai) [inline fairscale/optim/adascale.py:396]: perhaps rewrite this in a list comprehension way? Like:

[fn(x) for x in param_group["params"] for param_group in self._opt...]

in python, list comprehension is much faster than for loops. You can see some test I had here:

https://github.com/min-xu-ai/py_perf/blob/master/list_flatten/py_test.py

File: fairscale/optim/adascale.py [hunk 4]
```diff
@@ -387,6 +387,44 @@ def _update_avg(self, name: str, value: np.ndarray, factor: float) -> None:
...
+                elif p.grad.is_sparse:
+                    view = p.grad.to_dense().view(-1)
+                else:
+                    view = p.grad.view(-1)
+                views.append(view)
+        return torch.cat(views, 0)
+
+    def _compute_intra_grad_corr_mean(self) -> float:
+        """
+        Helper function for computing average intra correlation among gradients on different GPUs.
```
Comment (min-xu-ai) [inline fairscale/optim/adascale.py:409]: here we assume no_sync context is used, right? Perhaps document what is the requirement precisely here?

File: fairscale/optim/adascale.py [hunk 5]
```diff
@@ -387,6 +387,44 @@ def _update_avg(self, name: str, value: np.ndarray, factor: float) -> None:
...
+            dist.gather(flat_grad, gather_list=gathered_tensors, dst=0)
+            # the following requires torch 1.10+
+            corr = torch.stack(gathered_tensors).corrcoef()  # type: ignore
+            # pick out the upper triangular part of the correlation matrix
+            corr = corr[torch.triu(torch.ones_like(corr), diagonal=1) == 1]
+            corr_mean = corr.mean()
+        else:
+            dist.gather(flat_grad, gather_list=None, dst=0)
+        dist.broadcast(corr_mean, src=0)
+        return corr_mean.item()
```
Comment (min-xu-ai) [inline fairscale/optim/adascale.py:426]: .item() call would introduce a cpu-gpu sync. Is it actually needed here? I feel that it should be the caller does it in case it needs the float value. Otherwise, just let this function return the tensor to allow async execution of the cpu thread is better.

File: fairscale/optim/adascale.py [hunk 6]
```diff
@@ -449,6 +487,9 @@ def _final_callback(self) -> None:
             return
 
         # Since self._local_grad_sqr is FP32, sum shouldn't overflow.
+
+        # TODO: Hongbo says param.grad might be FP16 should do this before converting to FP32.
```
Comment (min-xu-ai) [inline fairscale/optim/adascale.py:491]: this is a bit too sparse. Maybe you can expand this and provide more context?

File: requirements-dev.txt [hunk 7]
```diff
@@ -32,7 +32,7 @@ pynvml == 8.0.4
 numpy == 1.22.0
 
 # For layerwise gradient scaler
-sklearn >= 0.0
+scikit-learn >= 0.0
```
Comment (min-xu-ai) [inline requirements-dev.txt:35]: nice! please use == to ensure it is not changing in the future when a new release is made by scikit-learn.

**Review by min-xu-ai — COMMENTED** [review]
Thanks for working on this! I left some comments. If you need help on the CI errors, let me know. The code is in general very clean.

File: .circleci/config.yml (hunk 2 above)
Comment (cyugao) [reply .circleci/config.yml:108]: For stable, I will do
`pip install --progress-bar off torch==1.13.0 torchvision==0.14.0`.
For nightly versions, I only see builds for cu116 or cu117. That's fine?

File: fairscale/optim/adascale.py (hunk 3 above)
Comment (min-xu-ai) [reply fairscale/optim/adascale.py:396]: I am just a perf improvement. no logical changes in the code
